freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/WorkloadCredentialService.java

```diff
@@ -29,46 +31,62 @@
     @Inject
     private BatchPartitionSizeProperties batchPartitionSizeProperties;
 
-    public void setWorkloadCredential(FreeIpaClient freeIpaClient, String username, WorkloadCredential workloadCredential)
+    @Inject
+    private UserMetadataConverter userMetadataConverter;
+
+    public void setWorkloadCredential(boolean credentialsUpdateOptimizationEnabled, FreeIpaClient freeIpaClient, WorkloadCredentialUpdate update)
             throws IOException, FreeIpaClientException {
-        LOGGER.debug("Setting workload credentials for user '{}'", username);
-        getOperation(username, workloadCredential, freeIpaClient).invoke(freeIpaClient);
+        LOGGER.debug("Setting workload credentials for user '{}'", update.getUsername());
+        getOperation(update.getUsername(), update.getUserCrn(), update.getWorkloadCredential(), credentialsUpdateOptimizationEnabled, freeIpaClient)
+                .invoke(freeIpaClient);
     }
 
-    public void setWorkloadCredentials(boolean fmsToFreeipaBatchCallEnabled, FreeIpaClient freeIpaClient, Map<String, WorkloadCredential> workloadCredentials,
+    public void setWorkloadCredentials(UserSyncOptions options, FreeIpaClient freeIpaClient, Set<WorkloadCredentialUpdate> updates,
```

Comment:
could you check this refactor. it should do the same but I think it's more readable:
```
    public void setWorkloadCredentials(UserSyncOptions options, FreeIpaClient freeIpaClient, Set<WorkloadCredentialUpdate> updates,
            BiConsumer<String, String> warnings) throws FreeIpaClientException {
        if (!updates.isEmpty()) {
            List<SetWlCredentialOperation> operations = updates.stream()
                    .flatMap(update -> createUpdateOperation(options, freeIpaClient, warnings, update))
                    .collect(Collectors.toList());
            if (options.fmsToFreeIpaBatchCallEnabled()) {
                updateInBatch(freeIpaClient, warnings, operations);
            } else {
                updateOneByOne(freeIpaClient, warnings, operations);
            }
        } else {
            LOGGER.debug("Nothing to update");
        }
    }

    private void updateOneByOne(FreeIpaClient freeIpaClient, BiConsumer<String, String> warnings, List<SetWlCredentialOperation> operations) throws FreeIpaClientException {
        LOGGER.debug("updateOneByOne log");
        for (SetWlCredentialOperation operation : operations) {
            try {
                operation.invoke(freeIpaClient);
            } catch (FreeIpaClientException e) {
                recordWarning(operation.getUser(), e, warnings);
                freeIpaClient.checkIfClientStillUsable(e);
            }
        }
    }

    private void updateInBatch(FreeIpaClient freeIpaClient, BiConsumer<String, String> warnings, List<SetWlCredentialOperation> operations) throws FreeIpaClientException {
        LOGGER.debug("Some log");
        List<Object> batchCallOperations = operations.stream()
                .map(AbstractFreeipaOperation::getOperationParamsForBatchCall)
                .collect(Collectors.toList());
        String operationName = operations.stream()
                .map(UserModOperation::getOperationName)
                .findFirst().orElse("unknown");
        freeIpaClient.callBatch(warnings, batchCallOperations, batchPartitionSizeProperties.getByOperation(operationName), Set.of());
    }

    private Stream<SetWlCredentialOperation> createUpdateOperation(UserSyncOptions options, FreeIpaClient freeIpaClient, BiConsumer<String, String> warnings, WorkloadCredentialUpdate update) {
        try {
            return Stream.of(getOperation(update.getUsername(), update.getUserCrn(), update.getWorkloadCredential(),
                    options.credentialsUpdateOptimizationEnabled(), freeIpaClient));
        } catch (IOException e) {
            recordWarning(update.getUsername(), e, warnings);
            LOGGER.warn("I would log this here");
            return Stream.empty();
        }
    }

Reply:
Done.